pkg/operator/terminationobserver/termination_observer.go

```diff
@@ -0,0 +1,272 @@
+		"TerminationPreShutdownHooksFinished",
+		"TerminationPreShutdownHooksFinished",
+		"TerminationMinimalShutdownDurationFinished",
+		"TerminationStoppedServing",
+		"TerminationGracefulTerminationFinished",
+	}
+)
+
+type TerminationObserver struct {
+	targetNamespace string
+
+	podsGetter corev1client.PodsGetter
+
+	cachesToSync  []cache.InformerSynced
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+
+	apiServers      []*apiServerStatus
+	apiServersMutex sync.RWMutex
+}
+
+type apiServerStatus struct {
+	serverName string
+	// currentCreatedTimestamp is creation timestamp of currently running API server
+	// this API server does not have any termination events, those events are only valid for
+	// "old" API server.
+	currentCreatedTimestamp time.Time
+	// terminatedTimestamp is timestamp that represent "old" API server that is currently being terminated.
+	// this timestamp can be nil if the API server was never terminated.
+	terminatedTimestamp *time.Time
+	terminationEvents   []*corev1.Event
+}
+
+func NewTerminationObserver(
+	targetNamespace string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	podsGetter corev1client.PodsGetter,
+	eventRecorder events.Recorder,
+) *TerminationObserver {
+	c := &TerminationObserver{
+		targetNamespace: targetNamespace,
+		podsGetter:      podsGetter,
+		eventRecorder:   eventRecorder.WithComponentSuffix("termination-observer"),
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TerminationObserver"),
+	}
+
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Events().Informer().AddEventHandler(c.terminationEventRecorder())
+
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Pods().Informer().HasSynced)
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Events().Informer().HasSynced)
+
+	return c
+}
+
+func (c *TerminationObserver) Register() {
+	registerMetrics(c.getApiServers)
+}
+
+func (c *TerminationObserver) getApiServers() []*apiServerStatus {
+	c.apiServersMutex.RLock()
+	defer c.apiServersMutex.RUnlock()
+	return c.apiServers
+}
+
+// updateApiServerCreationTimestamp updates the creation timestamp for single API server
+func (c *TerminationObserver) updateApiServerCreationTimestamp(name string, currentCreationTimestamp time.Time, isNew bool) {
+	c.apiServersMutex.Lock()
+	defer c.apiServersMutex.Unlock()
+	found := false
+	for i := range c.apiServers {
+		if c.apiServers[i].serverName != name {
+			continue
+		}
+		found = true
+		klog.Infof("Observed API server pod %q termination at %s", c.apiServers[i].serverName, currentCreationTimestamp)
+		if !isNew {
+			c.apiServers[i].terminatedTimestamp = &c.apiServers[i].currentCreatedTimestamp
+		}
+		c.apiServers[i].currentCreatedTimestamp = currentCreationTimestamp
+		break
+	}
+	if !found {
```

Comment:
why `isNew` and `found` logic? Isn't that the inverse?

Reply:
it is tricky... at the time when the creationTimestamp is updated on pod, the events you are receiving are for the "previous" pod that is now terminating... is `isNew` means this api server is "new" and was not running before (new master node?)